### DISCUSSION TITLE
why don't random() and randomSeed() conform to Arduino specifications?

### DISCUSSION THREAD

**Discussion by rsiemens77** [espressif/arduino-esp32#7399] [Q&A] [answered]
According to the Arduino API reference

https://www.arduino.cc/reference/en/language/functions/random-numbers/randomseed/

it should be possible to create a repeatable sequence of pseudo-random numbers based on the same seed being used.

This does not seem to work on ESP32.  Poking around, it looks like it is implemented here

https://github.com/espressif/arduino-esp32/blob/5be3ff74ea6057c7ee35fcc0cd60defb82786ec3/cores/esp32/WMath.cpp

It looks like it is based on esp_random() which uses noise to more "truly" randomize numbers.  This might seem like a good idea, but it actually breaks documented behavior.  The spec even suggests:

"it can occasionally be useful to use pseudo-random sequences that repeat exactly."

I think it would be better to keep the defined Arduino behavior and give the user a way to opt for the ESP32 style randomization if desired, or perhaps documenting a technique to do so, such as setting a seed based on esp_random().

**Comment by igrr** [comment, marked as answer]
Well spotted @rsiemens77!

This issue also used to exist in the ESP8266 Arduino core. On one hand, some libraries in the Arduino ecosystem used the `random` function inappropriately to generate some secrets or keys, hence it was desirable to make the source of these numbers truly random. On the other hand, there was the concern you have mentioned — sometimes it's desirable to have a repeatable pseudorandom sequence, for example in games.

The solution which was implemented in the ESP8266 Arduino core was to use the HW RNG unless `randomSeed` function was called. This was based on the observation that libraries didn't usually call `randomSeed`, and only called `random`. So calling `randomSeed` was an indication that the sketch author explicitly requests a random sequence based on a particular value.

Relevant links:
* https://github.com/esp8266/Arduino/issues/1710
* https://github.com/esp8266/Arduino/pull/2142

I think the same approach could be applied in the arduino-esp32 project.

**Reply by rsiemens77** [reply]
Thanks for the reference.  Nice that it was already discussed and resolved, so at least you won't have to go through that again.

On further investigation, there's another oddity in the current implementation. randomSeed() has no relevance to its API (ie. to random()), but it is actually invoking the C api srand() as though someone might want that.  This would then have the odd side-effect of breaking defined C api behavior if someone were using rand() and srand().

For anyone looking for a workaround to get a repeatable sequence for now, just use the C api's srand() and rand().

I'm surprised this hasn't been noticed: a guaranteed repeatable sequence is used and required in a variety of scenarios, which is exactly why this is a feature, and not undesirable, as some seem to naively assume.

**Comment by dsyleixa** [comment]
in general, always the highest possible Arduino-code-compatibilty is needed, in any respect.
Absolutely provided cross-board code-compatibilty is the crucial requirement for all board-specific API libs.
